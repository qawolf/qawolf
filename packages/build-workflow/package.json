{
  "name": "@qawolf/build-workflow",
  "description": "build qawolf workflow from events",
  "version": "0.9.0",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@qawolf/browser": "file:../browser",
    "@qawolf/logger": "file:../logger",
    "@qawolf/web": "file:../web"
  },
  "devDependencies": {
    "@qawolf/test": "file:../test",
    "@qawolf/types": "file:../types"
  },
  "peerDependencies": {
    "lodash": "^4.x"
  }
}
