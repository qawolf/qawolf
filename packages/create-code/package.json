{
  "name": "@qawolf/create-code",
  "description": "qawolf: create code",
  "version": "0.9.0",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@qawolf/build-code": "file:../build-code",
    "@qawolf/build-workflow": "file:../build-workflow",
    "@qawolf/logger": "file:../logger",
    "@qawolf/repl": "file:../repl"
  },
  "devDependencies": {
    "@qawolf/test": "file:../test",
    "@qawolf/types": "file:../types"
  },
  "peerDependencies": {
    "fs-extra": "^8.x",
    "kleur": "^3.x",
    "lodash": "^4.x"
  }
}
