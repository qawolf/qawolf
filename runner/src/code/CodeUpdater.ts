import Debug from "debug";
import { EventEmitter } from "events";
import { omit } from "lodash";
import { BrowserContext } from "playwright";

import { CodeUpdate, ElementEvent, Variables, WindowEvent } from "../types";
import { ContextEventCollector } from "./ContextEventCollector";
import { parseActionExpressions } from "./parseCode";
import { PATCH_HANDLE } from "./patch";
import { patchEvent, PatchEventOptions } from "./patchEvent";
import { patchFill, PatchFillOptions } from "./patchFill";
import { patchPopup, PatchPopupOptions } from "./patchPopup";
import { patchReload } from "./patchReload";

const debug = Debug("qawolf:CodeUpdater");

export const updateCode = (options: PatchEventOptions): string | null => {
  const { code, event } = options;
  const patchIndex = code.indexOf(PATCH_HANDLE);
  if (patchIndex < 0) return null;

  if (event.action === "fill") return patchFill(options as PatchFillOptions);
  if (event.action === "popup") return patchPopup(options as PatchPopupOptions);
  if (event.action === "reload") return patchReload(options);

  return patchEvent(options);
};

export class CodeUpdater extends EventEmitter {
  _code = "";
  _collector?: ContextEventCollector;
  _context?: BrowserContext;
  _enabled = false;
  _testId: string | undefined = "";
  _variables: Variables;
  _version = -1;

  constructor(variables: Variables) {
    super();
    this._variables = variables;
  }

  async _collectEvents(): Promise<void> {
    if (this._collector || !this._enabled || !this._context) return;

    this._collector = await ContextEventCollector.create(this._context);
    this._collector.on("elementevent", (event) => this._handleEvent(event));
    this._collector.on("windowevent", (event) => this._handleEvent(event));
  }

  _handleEvent(event: ElementEvent | WindowEvent): void {
    if (!this._enabled) return;

    debug("handle page event %o", omit(event, "frame", "page"));

    const updatedCode = updateCode({
      code: this._code,
      event,
      expressions: parseActionExpressions(this._code),
      variables: this._variables,
    });

    if (!updatedCode) {
      debug(`skip update: no changes`);
      return;
    }

    this._code = updatedCode;
    this._version += 1;

    const update: CodeUpdate = {
      code: this._code,
      generated: true,
      test_id: this._testId,
      version: this._version,
    };
    this.emit("codeupdated", update);
  }

  disable(): void {
    debug("disable");
    this._enabled = false;
  }

  async enable(): Promise<void> {
    debug("enable");
    this._enabled = true;
    await this._collectEvents();
  }

  async setContext(context: BrowserContext): Promise<void> {
    this._context = context;
    await this._collectEvents();
  }

  // Called when code is updated by the user
  updateCode(update: CodeUpdate): boolean {
    // Ignore updates that are lower versions that come from a race from multiple connected clients.
    // Allow the update if the test id changes which can happen when running locally and switching tests.
    // In the future we should reset everything when the test changes and move the test_id check.
    if (update.version <= this._version && update.test_id === this._testId) {
      return false;
    }

    debug(`update code ${update.version}`);
    this._code = update.code;
    this._testId = update.test_id;
    this._version = update.version;
    return true;
  }
}
