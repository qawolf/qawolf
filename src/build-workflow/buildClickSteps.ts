import Debug from 'debug';
import { ElementEvent, Step } from '../types';

const debug = Debug('qawolf:buildClickSteps');

const filterClickEvents = (events: ElementEvent[]): ElementEvent[] => {
  return events.filter((event, i) => {
    // track original event index
    (event as any).index = i;

    // ignore system initiated clicks
    if (!event.isTrusted) return false;

    // ignore other actions
    if (!['click', 'mousedown'].includes(event.name)) return false;

    const previousEvent = events[i - 1];
    if (
      previousEvent &&
      ['keydown', 'keyup'].includes(previousEvent.name) &&
      event.time - previousEvent.time < 50
    ) {
      // skip system-initiated clicks triggered by a key press
      // ex. "Enter" triggers a click on a submit input
      debug(`skip click shortly after previous event ${event.time}`);
      return false;
    }

    // ignore clicks on selects
    let name = event.target.name;
    if (name) name = name.toLowerCase();
    return name !== 'select';
  });
};

const groupClickEvents = (
  events: ElementEvent[],
  timeWindow = 200,
): ElementEvent[][] => {
  const groupedEvents = [];
  let group: ElementEvent[] = [];

  events.forEach(event => {
    if (!group.length) {
      // first group
      group.push(event);
      return;
    }

    const lastEvent = group[group.length - 1] as ElementEvent;
    if (event.time - lastEvent.time < timeWindow) {
      // event is within time window
      group.push(event);
      return;
    }

    // start a new group
    groupedEvents.push(group);
    group = [event];
  });

  if (group.length) {
    // append last group
    groupedEvents.push(group);
  }

  return groupedEvents;
};

export const buildClickSteps = (events: ElementEvent[]): Step[] => {
  const clickEvents = filterClickEvents(events);
  const groupedClickEvents = groupClickEvents(clickEvents);
  const steps: Step[] = [];

  groupedClickEvents.forEach(events => {
    let event = events[0] as ElementEvent;

    const inputEvent = events.find(event => {
      const name = event.target.name || '';
      return name.toLowerCase() === 'input';
    });
    if (inputEvent) {
      // if an event in the group is on an  input, assume the click propagated
      // to an element like a checkbox or radio, which is most accurate target
      event = inputEvent;
    }

    steps.push({
      action: 'click',
      index: steps.length,
      event: event,
    });
  });

  return steps;
};
