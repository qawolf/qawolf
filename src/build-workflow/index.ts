export * from './buildClickSteps';
export * from './buildKeySteps';
export * from './buildScrollSteps';
export * from './buildSelectSteps';
export * from './buildSteps';
