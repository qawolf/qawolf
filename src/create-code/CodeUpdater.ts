import Debug from 'debug';
import { buildVirtualCode } from '../build-code/buildVirtualCode';
import { CodeReconciler } from './CodeReconciler';
import { getLineIncludes, removeLinesIncluding } from './format';
import { PATCH_HANDLE } from './patchCode';
import { Step } from '../types';

const debug = Debug('qawolf:CodeUpdater');

export type CodeFileOptions = {
  path: string;
};

type UpdateOptions = {
  steps: Step[];
};

export const CREATE_HANDLE = `await qawolf.create`;

export abstract class CodeUpdater {
  protected _locked = false;
  private _reconciler: CodeReconciler;

  protected constructor() {
    this._reconciler = new CodeReconciler();
  }

  protected abstract async _loadCode(): Promise<string>;
  protected abstract async _updateCode(code: string): Promise<void>;

  protected async _prepare(): Promise<void> {
    this._locked = true;
    const code = await this._loadCode();

    const createLine = getLineIncludes(code, CREATE_HANDLE);
    if (!createLine) {
      throw new Error(`Could not find "${CREATE_HANDLE}"`);
    }

    // trim to match indentation
    const updatedCode = code.replace(createLine.trim(), PATCH_HANDLE);

    await this._updateCode(updatedCode);

    this._locked = false;
  }

  public async finalize(): Promise<void> {
    this._locked = true;
    let code = await this._loadCode();
    code = removeLinesIncluding(code, PATCH_HANDLE);
    await this._updateCode(code);
  }

  public async update(options: UpdateOptions): Promise<void> {
    // do not conflict with an update in progress
    if (this._locked) {
      debug(`skip update: update in progress`);
      return;
    }

    // check the virtual code changed
    const updatedVirtualCode = buildVirtualCode(options.steps);
    if (!this._reconciler.hasChanges(updatedVirtualCode)) {
      debug(`skip update: no virtual changes`);
      return;
    }

    this._locked = true;

    // update the actual code
    const actualCode = await this._loadCode();

    const updatedCode = this._reconciler.reconcile({
      actualCode,
      virtualCode: updatedVirtualCode,
    });
    await this._updateCode(updatedCode);

    // store the updated virtual code
    this._reconciler.update(updatedVirtualCode);

    this._locked = false;
  }
}
