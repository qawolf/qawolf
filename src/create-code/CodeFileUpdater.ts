import Debug from 'debug';
import { pathExists, readFile, remove, outputFile } from 'fs-extra';
import { CodeUpdater } from './CodeUpdater';

const debug = Debug('qawolf:CodeFileUpdater');

const loadCodeFile = async (path: string): Promise<string> => {
  const codeExists = await pathExists(path);
  if (!codeExists) throw new Error(`No code found at ${path}`);
  return readFile(path, 'utf8');
};

export class CodeFileUpdater extends CodeUpdater {
  private _path: string;
  protected _initialCode: string;

  public static async create(path: string): Promise<CodeFileUpdater> {
    debug(`load code from ${path}`);
    const initialCode = await loadCodeFile(path);
    const updater = new CodeFileUpdater(path);
    await updater._prepare();
    updater._initialCode = initialCode;
    return updater;
  }

  protected constructor(path: string) {
    super();
    this._path = path;
  }

  protected async _loadCode(): Promise<string> {
    return loadCodeFile(this._path);
  }

  protected async _updateCode(code: string): Promise<void> {
    await outputFile(this._path, code, 'utf8');
  }

  public async discard(): Promise<void> {
    this._locked = true;

    if (process.env.QAW_DISCARD === '1') {
      debug('discard code');
      await remove(this._path);
    } else {
      debug('revert to initial code');
      await this._updateCode(this._initialCode);
    }
  }

  public path(): string {
    return this._path;
  }
}
