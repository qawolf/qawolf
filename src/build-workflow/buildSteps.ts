import { concat, sortBy } from 'lodash';
import { buildClickSteps } from './buildClickSteps';
import { buildKeySteps } from './buildKeySteps';
import { buildScrollSteps } from './buildScrollSteps';
import { buildSelectSteps } from './buildSelectSteps';
import { ElementEvent, Step } from '../types';

export const buildSteps = (events: ElementEvent[], startIndex = 0): Step[] => {
  const unorderedSteps = concat(
    buildClickSteps(events),
    buildKeySteps(events),
    buildScrollSteps(events),
    buildSelectSteps(events),
  );

  let steps = sortBy(
    unorderedSteps,
    // ordered by the event time
    step => step.event.time,
  );

  // reindex
  steps = steps.map((step, index) => ({
    ...step,
    index: index + startIndex,
  }));

  return steps;
};
