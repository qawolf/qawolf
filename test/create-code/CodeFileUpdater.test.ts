import { outputFile, pathExists, readFile, remove } from 'fs-extra';
import { buildSteps } from '../../src/build-workflow/buildSteps';
import { CodeFileUpdater } from '../../src/create-code/CodeFileUpdater';
import { CREATE_HANDLE } from '../../src/create-code/CodeUpdater';

// require manually since fs is mocked
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { events } = require('../fixtures/login.json');
const steps = buildSteps(events);

jest.mock('fs-extra');

(pathExists as jest.Mock).mockResolvedValue(true);

const mockedReadFile = readFile as jest.Mock;
const mockedRemove = remove as jest.Mock;
const mockedOutputFile = outputFile as jest.Mock;

describe('CodeFileUpdater', () => {
  beforeEach(() => {
    mockedOutputFile.mockReset();
  });

  describe('discard', () => {
    it('reverts to initial code', async () => {
      const initialCode = `initialCode()\n${CREATE_HANDLE}`;
      mockedReadFile.mockResolvedValue(initialCode);
      const updater = await CodeFileUpdater.create('revertpath');
      await updater.discard();

      const reverted = mockedOutputFile.mock.calls[1];
      expect(reverted[0]).toEqual('revertpath');
      expect(reverted[1]).toEqual(initialCode);
      expect(mockedRemove.mock.calls.length).toEqual(0);
    });

    it('removes code if QAW_DISCARD is set to true', async () => {
      process.env.QAW_DISCARD = '1';
      const initialCode = `initialCode()\n${CREATE_HANDLE}`;
      mockedReadFile.mockResolvedValue(initialCode);
      const updater = await CodeFileUpdater.create('removepath');
      await updater.discard();
      expect(mockedRemove.mock.calls[0][0]).toEqual('removepath');
      process.env.QAW_DISCARD = undefined;
    });
  });

  describe('create', () => {
    it('replaces qawolf.create call with the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `  someCode();\n  await qawolf.create();\n  otherCode();`,
      );
      await CodeFileUpdater.create('somepath');

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe(
        `  someCode();\n  // 🐺 CREATE CODE HERE\n  otherCode();`,
      );
    });

    it('throws an error if qawolf.create is not found', async () => {
      mockedReadFile.mockResolvedValue('no create');
      // eslint-disable-next-line jest/valid-expect
      expect(CodeFileUpdater.create('somepath')).rejects.toThrowError();
    });
  });

  describe('finalize', () => {
    it('removes the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `someCode();\n${CREATE_HANDLE}\notherCode();`,
      );
      const updater = await CodeFileUpdater.create('somepath');
      const preparedCode = mockedOutputFile.mock.calls[0][1];
      mockedReadFile.mockResolvedValue(preparedCode);

      mockedOutputFile.mockReset();
      await updater.finalize();

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe('someCode();\notherCode();');
    });
  });

  describe('update', () => {
    it('saves code with new steps', async () => {
      mockedReadFile.mockResolvedValue(
        `  someCode();\n  ${CREATE_HANDLE}\n  otherCode();`,
      );

      const updater = await CodeFileUpdater.create('somepath');
      const preparedCode = mockedOutputFile.mock.calls[0][1];
      mockedReadFile.mockResolvedValue(preparedCode);

      // emulate only the first two steps being ready
      await updater.update({ steps: steps.slice(0, 2) });

      const codeUpdateOne = mockedOutputFile.mock.calls[1][1];
      expect(codeUpdateOne).toMatchSnapshot();
      mockedReadFile.mockResolvedValueOnce(codeUpdateOne);

      await updater.update({ steps });
      const codeUpdateTwo = mockedOutputFile.mock.calls[2][1];
      expect(codeUpdateTwo).toMatchSnapshot();
    });
  });
});
