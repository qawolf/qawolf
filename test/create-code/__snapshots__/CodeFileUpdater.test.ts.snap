// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeFileUpdater update saves code with new steps 1`] = `
"  someCode();
  await page.click(selectors[\\"0_a\\"]);
  await page.click(selectors[\\"1_username_input\\"]);
  // üê∫ CREATE CODE HERE
  otherCode();"
`;

exports[`CodeFileUpdater update saves code with new steps 2`] = `
"  someCode();
  await page.click(selectors[\\"0_a\\"]);
  await page.click(selectors[\\"1_username_input\\"]);
  await page.type(selectors[\\"2_username_input\\"], \\"tomsmith\\");
  await page.press(selectors[\\"3_username_input\\"], \\"Tab\\");
  await page.type(selectors[\\"4_password_input\\"], \\"SuperSecretPassword!\\");
  await page.press(selectors[\\"5_password_input\\"], \\"Enter\\");
  await page.click(selectors[\\"6_p\\"]);
  // üê∫ CREATE CODE HERE
  otherCode();"
`;
